keras_nlp/samplers/beam_sampler.py

```diff
@@ -39,192 +36,137 @@ class BeamSampler(Sampler):
-    def token_probability_fn(inputs, mask):
-        return model(inputs)
-
-    prompt = tf.fill((8, 1), 1)
-
-    sampler = keras_nlp.samplers.BeamSampler(num_beams=3)
-    # Print the generated sequence (token ids).
-    print(sampler(prompt, token_probability_fn, max_length=10))
+    print(["".join([int_lookup[i] for i in s]) for s in output.numpy()])
+    # >>> "zzzzzaaaaaaa"
     ```
     """
 
     def __init__(
         self,
         num_beams=5,
-        jit_compile=True,
-        run_eagerly=False,
     ):
+        super().__init__()
         self.num_beams = num_beams
-        super().__init__(jit_compile=jit_compile, run_eagerly=run_eagerly)
 
-    def get_next_token(self, next_token_probs):
-        # Beam search overrides the whole `sample` method.
-        pass
-
-    def sample(
+    def __call__(
         self,
+        next,
         prompt,
-        mask,
-        num_steps,
-        from_logits=True,
+        state=None,
+        index=0,
+        mask=None,
         end_token_id=None,
-        cache=None,
     ):
-        """Sampling logic implementation.
-
-        Because beam search uses a different loop body, we have to override the
-        whole `sample` method instead of just the `get_next_token` method.
-        """
-        if cache is not None:
-            logging.warning(
-                "`BeamSampler` does not support cache decoding now, the cache "
-                "will be ignored. To use cache decoding, please use a "
-                "different sampler, e.g., `TopPSampler`."
-            )
         batch_size, max_length = tf.shape(prompt)[0], tf.shape(prompt)[1]
-        max_length = tf.cast(max_length, num_steps.dtype)
-        length = max_length - num_steps
-        dummy_preds = self.token_probability_fn(prompt, mask=mask)
-        vocab_size = tf.shape(dummy_preds)[-1]
-        pred_dtype = dummy_preds.dtype
-
-        num_beams = self.num_beams
-
-        # Initialize beam with shape `(batch_size, num_beams, length)`.
-        beams = tf.repeat(tf.expand_dims(prompt, axis=1), num_beams, axis=1)
-        # Initialize `beams_prob` with shape `(batch_size, num_beams)`.
-        beams_prob = tf.zeros([batch_size, 1], dtype=pred_dtype)
-        beams_prob = tf.concat(
-            [beams_prob, tf.fill((batch_size, num_beams - 1), pred_dtype.min)],
-            axis=-1,
-        )
 
-        def one_step(beams, beams_prob, length, mask):
-            flattened_beams = tf.reshape(
-                beams, shape=[batch_size * num_beams, -1]
-            )
-            repeated_mask = tf.tile(mask, [num_beams, 1])
-            probs = self.token_probability_fn(flattened_beams, repeated_mask)
-            preds = tf.gather(
-                probs,
-                tf.repeat(length - 1, batch_size * num_beams),
-                axis=1,
-                batch_dims=1,
-            )
-            if from_logits:
-                preds = keras.activations.softmax(preds, axis=-1)
+        def add_beams(x):
```

Comment:
Optional renaming: to me `add_beams` suggest it could be called for multiple times, shall we go for `create_beams` or `build_beams` since it's only called before the generation loop?